nuxeo-coldstorage/src/main/java/org/nuxeo/coldstorage/service/ColdStorageServiceImpl.java

```diff
@@ -303,47 +320,65 @@ public DocumentModel requestRetrievalFromColdStorage(CoreSession session, Docume
-            Blob coldContent = (Blob) documentModel.getPropertyValue(COLD_STORAGE_CONTENT_PROPERTY);
-            String key = getContentBlobKey(coldContent);
-            BlobUpdateContext updateContext = new BlobUpdateContext(key).withRestoreForDuration(restoreDuration);
-            Framework.getService(BlobManager.class).getBlobProvider(coldContent).updateBlob(updateContext);
-        } catch (IOException e) {
-            throw new NuxeoException(e);
+        BlobStatus blobStatus = ColdStorageHelper.getBlobStatus(documentModel);
+        boolean notify = false;
+        DocumentModel docResult = null;
+        if (ColdStorageHelper.isDownloadable(blobStatus)) {
+            documentModel.setPropertyValue(COLD_STORAGE_CONTENT_DOWNLOADABLE_UNTIL,
+                    Date.from(blobStatus.getDownloadableUntil()));
+            docResult = CoreInstance.doPrivileged(session, s -> {
+                // The retrieval is allowed for users with only READ access.
+                // It requires an unrestricted session to update ColdStorage metadata on the document
+                return s.saveDocument(documentModel);
+            });
+        } else if (blobStatus.isOngoingRestore()) {
+            documentModel.setPropertyValue(COLD_STORAGE_BEING_RETRIEVED_PROPERTY, true);
+            notify = true;
+            docResult = CoreInstance.doPrivileged(session, s -> {
+                // The retrieval is allowed for users with only READ access.
+                // It requires an unrestricted session to update ColdStorage metadata on the document
+                return s.saveDocument(documentModel);
+            });
+        } else {
+            try {
+                Blob coldContent = (Blob) documentModel.getPropertyValue(COLD_STORAGE_CONTENT_PROPERTY);
+                String key = getContentBlobKey(coldContent);
+                BlobUpdateContext updateContext = new BlobUpdateContext(key).withRestoreForDuration(restoreDuration);
+                Framework.getService(BlobManager.class).getBlobProvider(coldContent).updateBlob(updateContext);
+            } catch (IOException e) {
+                log.error("Could not retrieve document {} for duration {} seconds", documentModel::getId,
+                        restoreDuration::getSeconds);
+                throw new NuxeoException(e);
+            }
+            documentModel.setPropertyValue(COLD_STORAGE_BEING_RETRIEVED_PROPERTY, true);
+            docResult = CoreInstance.doPrivileged(session, s -> {
+                // The retrieval is allowed for users with only READ access.
+                // It requires an unrestricted session to update ColdStorage metadata on the document
+                return s.saveDocument(documentModel);
+            });
+            notify = CoreInstance.doPrivileged(session, s -> {
+                // The check retrieval may need to modify metadata of document too
+                return !checkIsRetrieved(session, documentModel);
+            });
         }
```

Comment:
There's an additional thing to change on the condition later in the code, but I can't select it in the PR.
```suggestion
        Function<DocumentModel, Boolean> doNotify;
        if (ColdStorageHelper.isDownloadable(blobStatus)) {
            documentModel.setPropertyValue(COLD_STORAGE_CONTENT_DOWNLOADABLE_UNTIL,
                    Date.from(blobStatus.getDownloadableUntil()));
            doNotify = doc -> false;
        } else if (blobStatus.isOngoingRestore()) {
            documentModel.setPropertyValue(COLD_STORAGE_BEING_RETRIEVED_PROPERTY, true);
            doNotify = doc -> true;
        } else {
            try {
                Blob coldContent = (Blob) documentModel.getPropertyValue(COLD_STORAGE_CONTENT_PROPERTY);
                String key = getContentBlobKey(coldContent);
                BlobUpdateContext updateContext = new BlobUpdateContext(key).withRestoreForDuration(restoreDuration);
                Framework.getService(BlobManager.class).getBlobProvider(coldContent).updateBlob(updateContext);
            } catch (IOException e) {
                log.error("Could not retrieve document {} for duration {} seconds", documentModel::getId,
                        restoreDuration::getSeconds);
                throw new NuxeoException(e);
            }
            documentModel.setPropertyValue(COLD_STORAGE_BEING_RETRIEVED_PROPERTY, true);
            doNotify = doc -> CoreInstance.doPrivileged(session, s -> {
                // The check retrieval may need to modify metadata of document too
                return !checkIsRetrieved(s, doc);
            });
        }
        DocumentModel docResult = CoreInstance.doPrivileged(session, s -> {
            // The retrieval is allowed for users with only READ access.
            // It requires an unrestricted session to update ColdStorage metadata on the document
            return s.saveDocument(documentModel);
        });
```